### PR TITLE
add unit tests for machine scope methods

### PR DESCRIPTION
**Pull request by sonasingh46**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR adds unit test for machine scope methods as described in this issue: 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1108

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1108

**Special notes for your reviewer**:

Have added UT for the following methods: 
```
PublicIPSpecs()
InboundNatSpecs()
RoleAssignmentSpecs()
VMExtensionSpecs()
Subnet()
AvailabilityZone()
AvailabilitySet()
Role()
IsControlPlane()
Namespace()
VMState()
GetVMImage()
```
Will be adding couple of thus removing WIP 
_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by shysank** [comment]
/lgtm
Thanks @sonasingh46 , I think this will be super useful in catching errors early.

File: azure/scope/machine_test.go [hunk 1]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			want: infrav1.SubnetSpec{
+				Name: "machine-name-subnet",
+			},
+		},
+		{
+			name: "returns empty if machine subnet name is not present in the cluster scope",
```
Comment (mboersma) [inline azure/scope/machine_test.go:559]: This test spec looks the same as the one above it, but I think you meant to have it return an empty `SubnetSpec`.

File: azure/scope/machine_test.go [hunk 2]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+					},
+					Spec: infrav1.AzureMachineSpec{
+						FailureDomain: pointer.String("dummy-failure-domain-from-azuremachine-spec"),
+					},
+				},
+			},
+			want: "dummy-failure-domain-from-machine-spec",
+		},
+		{
+			name: "returns failure domain from the azuremachine spec",
```
Comment (CecileRobertMichon) [inline azure/scope/machine_test.go:641]: is this test case a duplicate from the one above?

File: azure/scope/machine_test.go [hunk 3]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "machine-name",
+						Namespace: "foo",
+					},
+				},
+			},
+			want: "foo",
+		},
+		{
+			name: "returns azure machine namespace#2",
```
Comment (mboersma) [inline azure/scope/machine_test.go:686]: Should this say `"returns empty azure machine namespace"`?

Comment (CecileRobertMichon) [inline azure/scope/machine_test.go:686]: can this test name be more descriptive?

File: azure/scope/machine_test.go [hunk 4]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+							provisioningState := infrav1.Creating
+							return &provisioningState
+						}(),
+					},
+				},
+			},
+			want: infrav1.Creating,
+		},
+		{
+			name: "returns empty if Vmstate is not present in AzureMachine status",
```
Comment (CecileRobertMichon) [inline azure/scope/machine_test.go:938]:
```suggestion
			name: "returns empty if VMState is not present in AzureMachine status",
```

File: azure/scope/machine_test.go [hunk 5]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+}
+
+func TestMachineScope_VMState(t *testing.T) {
+	tests := []struct {
+		name         string
+		machineScope MachineScope
+		want         infrav1.ProvisioningState
+	}{
+		{
+			name: "returns the VMstate if present in AzureMachine status",
```
Comment (CecileRobertMichon) [inline azure/scope/machine_test.go:921]:
```suggestion
			name: "returns the VMState if present in AzureMachine status",
```

File: azure/scope/machine_test.go [hunk 6]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+							clusterv1.MachineDeploymentLabelName: "foo-machine-deployment",
+						},
+					},
+				},
+			},
+			wantAvailabilitySetName:      "cluster_foo-machine-deployment-as",
+			wantAvailabilitySetExistence: true,
+		},
+		{
+			name: "returns empty and true if AvailabilitySet is enabled but worker machine is not part of machine deployment",
```
Comment (mboersma) [inline azure/scope/machine_test.go:878]:
```suggestion
			name: "returns empty and false if AvailabilitySet is enabled but worker machine is not part of machine deployment",
```

**Review by CecileRobertMichon — COMMENTED** [review]
looks good, added a few comments. thanks @sonasingh46 !

File: azure/scope/machine_test.go [hunk 7]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			want: &infrav1.Image{
+				ID: pointer.StringPtr("1"),
+			},
+			wantErr: false,
+		},
+
```
Comment (mboersma) [inline azure/scope/machine_test.go:985]:
```suggestion
```

File: azure/scope/machine_test.go [hunk 8]
```diff
@@ -223,3 +229,829 @@ func TestMachineScope_ProviderID(t *testing.T) {
...
+					},
+				},
+			},
+			want: func() *infrav1.Image {
+				image, _ := azure.GetDefaultWindowsImage("1.20.1")
+				return image
+			}(),
+			wantErr: false,
+		},
+
```
Comment (mboersma) [inline azure/scope/machine_test.go:1015]:
```suggestion
```

**Review by mboersma — CHANGES_REQUESTED** [review]

File: azure/scope/machine_test.go (hunk 2 above)
Comment (sonasingh46) [reply azure/scope/machine_test.go:641]: No. If failure domain is specified in both `MachineSpec` as well as `AzureMachineSpec` than the function returns the failure domain from `MachineSpec`. 
In this test case failure domain is not specified on `MachineSpec` and in above test case failure domain is specified on both `MachineSpec` as well as `AzureMachineSpec`

File: azure/scope/machine_test.go (hunk 1 above)
Comment (sonasingh46) [reply azure/scope/machine_test.go:559]: Yes, thank you. That is a slip from me. Will fix this.

Comment (sonasingh46) [reply azure/scope/machine_test.go:559]: fixed the test spec

**Review by mboersma — APPROVED** [review]
/lgtm

Adding unit tests is so appreciated! Thank you.

*Edit:* this still needs to be squashed into a single commit, btw.

**Review by mboersma — APPROVED** [review]

**Comment by CecileRobertMichon** [comment]
/lgtm
/approve

Thanks @sonasingh46 !

**Pull request merged**